src/deps-index.ts

```diff
@@ -81,28 +29,170 @@ function _getHtmlDependencies(
+  return subBundleUrl.replace(/>[^>]+$/, '') as ResolvedUrl;
+}
+
+type DependencyMapEntry = {
+  // All dependencies of the document
+  deps: Set<ResolvedUrl>,
+  // Eagerly loaded dependencies of the document
+  eagerDeps: Set<ResolvedUrl>,
+  // All imports defined with `<link rel="lazy-import">` or with dynamic ES
+  // module import syntax like `import().then()`
+  lazyImports: Set<ResolvedUrl>,
+  // All imports defined with `<script type="module" src="...">` or as import
+  // statements within a `<script type="module">...</script>`
+  moduleScriptImports: Map<string, Document>,
+};
+
+/**
+ * These are the options included in every `Document#getFeatures` call, DRY'd up
+ * here for brevity and consistency.
+ */
+const getFeaturesOptions = {
+  imported: false,
+  externalPackages: true,
+  excludeBackreferences: true,
+};
+
+/**
+ * For a given document, return a set of transitive dependencies, including
+ * all eagerly-loaded dependencies and lazy html imports encountered.
+ */
+function getDependencies(document: Document): DependencyMapEntry {
+  const deps = new Set<ResolvedUrl>();
+  const eagerDeps = new Set<ResolvedUrl>();
+  const lazyImports = new Set<ResolvedUrl>();
+  const moduleScriptImports = new Map<string, Document>();
+  _getDependencies(document, true);
+  return {deps, eagerDeps, lazyImports, moduleScriptImports};
+
+  function _getDependencies(document: Document, viaEager: boolean) {
+    // HTML document dependencies include external modules referenced by script
+    // src attribute, external modules imported by inline module import
+    // statements, and HTML imports (recursively).
+    if (document.kinds.has('html-document')) {
+      _getHtmlExternalModuleDependencies(document);
+      _getHtmlInlineModuleDependencies(document);
+      _getImportDependencies(
+          document.getFeatures({kind: 'html-import', ...getFeaturesOptions}),
+          viaEager);
+    }
+
+    // JavaScript documents, when parsed as modules, have dependencies defined
+    // by their import statements.
+    if (document.kinds.has('js-document')) {
+      _getImportDependencies(
+          [...document.getFeatures({kind: 'js-import', ...getFeaturesOptions})]
+              .filter(
+                  (d) => (d.document.parsedDocument as JavaScriptDocument)
+                             .parsedAsSourceType === 'module'),
+          viaEager);
+    }
+  }
+
+  function _getHtmlExternalModuleDependencies(document: Document) {
+    let externalModuleCount = 0;
+    const htmlScripts =
+        [...document.getFeatures({kind: 'html-script', ...getFeaturesOptions})]
+            .filter(
+                (i) => (i.document.parsedDocument as JavaScriptDocument)
+                           .parsedAsSourceType === 'module');
+    for (const htmlScript of htmlScripts) {
+      moduleScriptImports.set(
+          `external-module:${++externalModuleCount}`, htmlScript.document);
```

Comment:
how hard would it be to include information about the original `htmlScript.url` in this map key? This feels like useful info that shouldn't get lost, at the very least for debugging.

Reply:
Added!